### ISSUE TITLE
Crash on button press -> java.lang.RuntimeException: View must have a tag

### ISSUE THREAD

**Issue by grad1e** [androidbroadcast/ViewBindingPropertyDelegate#65]
Hi, I'm facing an issue after upgrading to 1.4.7 while clicking a button. Downgrading to 1.4.6 fixed for me. Here's the stack trace and code

### Stacktrace
```
E/AndroidRuntime: FATAL EXCEPTION: main
    Process: com.atees.ayurdoc.debug, PID: 25736
    java.lang.RuntimeException: java.lang.reflect.InvocationTargetException
        at com.android.internal.os.RuntimeInit$MethodAndArgsCaller.run(RuntimeInit.java:602)
        at com.android.internal.os.ZygoteInit.main(ZygoteInit.java:947)
     Caused by: java.lang.reflect.InvocationTargetException
        at java.lang.reflect.Method.invoke(Native Method)
        at com.android.internal.os.RuntimeInit$MethodAndArgsCaller.run(RuntimeInit.java:592)
        at com.android.internal.os.ZygoteInit.main(ZygoteInit.java:947) 
     Caused by: java.lang.reflect.InvocationTargetException
        at java.lang.reflect.Method.invoke(Native Method)
        at by.kirich1409.viewbindingdelegate.internal.BindViewBinding.bind(ViewBindingCache.kt:101)
        at by.kirich1409.viewbindingdelegate.ReflectionFragmentViewBindings$viewBinding$3.invoke(FragmentViewBindings.kt:64)
        at by.kirich1409.viewbindingdelegate.ReflectionFragmentViewBindings$viewBinding$3.invoke(FragmentViewBindings.kt:63)
        at by.kirich1409.viewbindingdelegate.LifecycleViewBindingProperty.getValue(ViewBindingProperty.kt:77)
        at by.kirich1409.viewbindingdelegate.FragmentViewBindingProperty.getValue(FragmentViewBindings.kt:42)
        at by.kirich1409.viewbindingdelegate.FragmentViewBindingProperty.getValue(FragmentViewBindings.kt:34)
        at com.atees.ayurdoc.ui.login.LoginFragment.getBinding(LoginFragment.kt:33)
        at com.atees.ayurdoc.ui.login.LoginFragment.setupClickListeners$lambda-4(LoginFragment.kt:124)
        at com.atees.ayurdoc.ui.login.LoginFragment.lambda$snF-fOR_eVlj8Zahf8h-q80glhw(Unknown Source:0)
        at com.atees.ayurdoc.ui.login.-$$Lambda$LoginFragment$snF-fOR_eVlj8Zahf8h-q80glhw.onClick(Unknown Source:2)
        at android.view.View.performClick(View.java:7448)
        at com.google.android.material.button.MaterialButton.performClick(MaterialButton.java:1119)
        at android.view.View.performClickInternal(View.java:7425)
        at android.view.View.access$3600(View.java:810)
        at android.view.View$PerformClick.run(View.java:28305)
        at android.os.Handler.handleCallback(Handler.java:938)
        at android.os.Handler.dispatchMessage(Handler.java:99)
        at android.os.Looper.loop(Looper.java:223)
        at android.app.ActivityThread.main(ActivityThread.java:7664)
        at java.lang.reflect.Method.invoke(Native Method) 
        at com.android.internal.os.RuntimeInit$MethodAndArgsCaller.run(RuntimeInit.java:592) 
        at com.android.internal.os.ZygoteInit.main(ZygoteInit.java:947) 
     Caused by: java.lang.RuntimeException: view must have a tag
        at com.atees.ayurdoc.DataBinderMapperImpl.getDataBinder(DataBinderMapperImpl.java:75)
        at androidx.databinding.MergedDataBinderMapper.getDataBinder(MergedDataBinderMapper.java:74)
        at androidx.databinding.DataBindingUtil.bind(DataBindingUtil.java:199)
        at androidx.databinding.ViewDataBinding.bind(ViewDataBinding.java:695)
        at com.atees.ayurdoc.databinding.FragmentLoginBinding.bind(FragmentLoginBinding.java:136)
        at com.atees.ayurdoc.databinding.FragmentLoginBinding.bind(FragmentLoginBinding.java:124)
        at java.lang.reflect.Method.invoke(Native Method) 
        at by.kirich1409.viewbindingdelegate.internal.BindViewBinding.bind(ViewBindingCache.kt:101) 
        at by.kirich1409.viewbindingdelegate.ReflectionFragmentViewBindings$viewBinding$3.invoke(FragmentViewBindings.kt:64) 
        at by.kirich1409.viewbindingdelegate.ReflectionFragmentViewBindings$viewBinding$3.invoke(FragmentViewBindings.kt:63) 
        at by.kirich1409.viewbindingdelegate.LifecycleViewBindingProperty.getValue(ViewBindingProperty.kt:77) 
        at by.kirich1409.viewbindingdelegate.FragmentViewBindingProperty.getValue(FragmentViewBindings.kt:42) 
        at by.kirich1409.viewbindingdelegate.FragmentViewBindingProperty.getValue(FragmentViewBindings.kt:34) 
        at com.atees.ayurdoc.ui.login.LoginFragment.getBinding(LoginFragment.kt:33) 
        at com.atees.ayurdoc.ui.login.LoginFragment.setupClickListeners$lambda-4(LoginFragment.kt:124) 
        at com.atees.ayurdoc.ui.login.LoginFragment.lambda$snF-fOR_eVlj8Zahf8h-q80glhw(Unknown Source:0) 
        at com.atees.ayurdoc.ui.login.-$$Lambda$LoginFragment$snF-fOR_eVlj8Zahf8h-q80glhw.onClick(Unknown Source:2) 
        at android.view.View.performClick(View.java:7448) 
        at com.google.android.material.button.MaterialButton.performClick(MaterialButton.java:1119) 
        at android.view.View.performClickInternal(View.java:7425) 
        at android.view.View.access$3600(View.java:810) 
        at android.view.View$PerformClick.run(View.java:28305) 
        at android.os.Handler.handleCallback(Handler.java:938) 
        at android.os.Handler.dispatchMessage(Handler.java:99) 
        at android.os.Looper.loop(Looper.java:223) 
        at android.app.ActivityThread.main(ActivityThread.java:7664) 
        at java.lang.reflect.Method.invoke(Native Method) 
        at com.android.internal.os.RuntimeInit$MethodAndArgsCaller.run(RuntimeInit.java:592) 
        at com.android.internal.os.ZygoteInit.main(ZygoteInit.java:947) 
```
### Fragment code
```
@AndroidEntryPoint
class LoginFragment : Fragment(R.layout.fragment_login) {

    private val binding: FragmentLoginBinding by viewBinding()
    private val viewModel: LoginViewModel by viewModels()
    private val args: LoginFragmentArgs by navArgs()

    override fun onViewCreated(view: View, savedInstanceState: Bundle?) {
        super.onViewCreated(view, savedInstanceState)
        binding.lifecycleOwner = viewLifecycleOwner
        binding.viewModel = viewModel
        setupClickListeners()
    }

    private fun setupClickListeners() {
        binding.btnLogin.setOnClickListener {
            binding.root.showSnackBar("Pressed")
        }
    }
}
```
### Extension function
```
fun View.showSnackBar(message: String) {
        Snackbar.make(this, message, Snackbar.LENGTH_LONG).show()
}
```

**Comment by kirich1409** [comment]
Please, create repo with sample where the issue is reproducible

**Comment by 1240** [comment]
same

**Comment by kirich1409** [comment]
> same

Please, create repo with sample where the issue is reproducible

**Comment by kirich1409** [comment]
Quick search give me that issues connected with DataBinding https://stackoverflow.com/questions/53184433/what-the-actual-meaning-caused-by-java-lang-runtimeexception-view-must-have-a

ViewBindingPropertyDelegate doesn't support it and I didn't test with DataBinding, but will do that

**Comment by kirich1409** [comment]
See #69

**Comment by kirich1409** [comment]
@daryljodanny show code of the layout resource file that crash when use it with VBPD

**Comment by daryljodanny** [comment]
I was so busy that I literally forgot everything. I’ll help you with this by today night for sure

**Comment by Berki2021** [comment]
Got the same error today, also using databinding. It is unnecessary, how the layout file looks like or if you use viewbindingdelegate with or without reflection. Always gives me this error. Here is my layout for research purpose:

### Fragment Layout with viewpager2

```
<layout xmlns:android="http://schemas.android.com/apk/res/android"
    xmlns:app="http://schemas.android.com/apk/res-auto">

    <androidx.constraintlayout.widget.ConstraintLayout
        android:layout_width="match_parent"
        android:layout_height="match_parent">

        <include
            android:id="@+id/headline"
            layout="@layout/registration_toolbar"
            app:isBold="@{false}"
            app:layout_constraintEnd_toEndOf="parent"
            app:layout_constraintStart_toStartOf="parent"
            app:layout_constraintTop_toTopOf="parent"
            app:outsideToolbarTitle="@{@string/fragment_user_data_toolbar_title}" />

        <com.google.android.material.tabs.TabLayout
            android:id="@+id/tl_user_data"
            android:layout_width="match_parent"
            android:layout_height="wrap_content"
            app:layout_constraintEnd_toEndOf="parent"
            app:layout_constraintStart_toStartOf="parent"
            app:layout_constraintTop_toBottomOf="@+id/headline" />

        <androidx.viewpager2.widget.ViewPager2
            android:id="@+id/vp_user_data"
            android:layout_width="match_parent"
            android:layout_height="0dp"
            android:layout_marginTop="8dp"
            app:layout_constraintBottom_toBottomOf="parent"
            app:layout_constraintEnd_toEndOf="parent"
            app:layout_constraintStart_toStartOf="parent"
            app:layout_constraintTop_toBottomOf="@+id/tl_user_data" />

    </androidx.constraintlayout.widget.ConstraintLayout>
</layout>
```

### Headline layout file

```
<layout>

    <data>

        <variable
            name="outsideToolbarTitle"
            type="String" />

        <variable
            name="isBold"
            type="boolean" />
    </data>

    <androidx.constraintlayout.widget.ConstraintLayout xmlns:android="http://schemas.android.com/apk/res/android"
        xmlns:app="http://schemas.android.com/apk/res-auto"
        android:layout_width="match_parent"
        android:layout_height="wrap_content">

        <com.google.android.material.appbar.MaterialToolbar
            android:id="@+id/user_standard_toolbar"
            android:layout_width="match_parent"
            android:layout_height="wrap_content"
            android:background="@color/color_shop_item_background"
            android:theme="@style/ThemeOverlay.AppCompat.Light"
            app:layout_constraintEnd_toEndOf="parent"
            app:layout_constraintStart_toStartOf="parent"
            app:layout_constraintTop_toTopOf="parent">

            <TextView
                android:id="@+id/toolbar_title"
                android:layout_width="wrap_content"
                android:layout_height="wrap_content"
                android:layout_gravity="center"
                android:text="@{outsideToolbarTitle}"
                android:textColor="@android:color/black"
                android:textSize="@dimen/textHeadlineNormal1"
                app:setStyle="@{isBold ? 1 : 0}" />
        </com.google.android.material.appbar.MaterialToolbar>

    </androidx.constraintlayout.widget.ConstraintLayout>
</layout>
```

**Comment by kirich1409** [comment]
I planed to add callback to the delegate, when view will be destroyed

**Comment by wbervoets** [comment]
Hello, is databinding support still being worked on? I tried with 1.5.2 but the crash is still happening. For me it is not related to onDestroyView like @Berki2021 but as the original poster: in onViewCreated when the viewBinding is accessed from a click listener on a button.

1.4.6 works fine, so something changed in 1.4.7 and later.

If you cannot reproduce, please let me know

**Comment by kirich1409** [comment]
> Hello, is databinding support still being worked on? I tried with 1.5.2 but the crash is still happening. For me it is not related to onDestroyView like @Berki2021 but as the original poster: in onViewCreated when the viewBinding is accessed from a click listener on a button.
> 
> 1.4.6 works fine, so something changed in 1.4.7 and later.
> 
> If you cannot reproduce, please let me know

It will be great if you provide sample where the issue is reproducible

**Comment by wbervoets** [comment]
I will try to create a minimal sample.

What I have observed so far is that 

`private val viewBinding: FragmentLoginScreenBinding by viewBinding(
        FragmentLoginScreenBinding::bind
    )`

works fine when viewBinding is accessed in onViewCreated() of the fragment.

It crashes when observing a livedata which is called in onViewCreated(). We pass in the viewLifecycleowner, not the fragment lifecycle owner.

`viewModel.login().observe(
                viewLifecycleOwner,
                {
                    when (it) {
                        is Result.Loading -> {
                            viewBinding.loginButton.showProgress()
                        }
                        is Result.Success -> {
}`

When  viewBinding.loginButton.showProgress() is executed it will crash.
This is due to the root ConstraintLayout of the login_fragment not having a tag anymore. (while it does have the correct tag in the xml in the apk)

I noticed the viewBinding became null again in your LifecycleViewBindingProperty, which seems strange:

`@MainThread
    public override fun getValue(thisRef: R, property: KProperty<*>): T {
        viewBinding?.let { return it }`

It seems to be cleared by ClearOnDestroyLifecycleObserver which is triggered when the view lifecycleowner is destroyed.
So why is it destroyed? 

First I see it being cleared on a previous fragment onDestroyView() is called (which contains a  viewBinding access).
But then the new login fragment which afaik is stil visible also gets a callback in ClearOnDestroyLifecycleObserver.

**Comment by wbervoets** [comment]
Some more information:

I have two fragments: Home and Login. Home is the first fragment that is shown, it automatically redirects to Login when the user is not logged in.

So the following happens:
- Home fragment onViewCreated called
- Navigate to Login fragment
- Login Fragment onViewCreated called
- Login fragment is shown
At this point viewbinding can be accessed and I don't receive any crashes due to "view must have a tag"

A bit later the following happens:
- Home Fragment onDestroyView is called
- This triggers the ClearOnDestroy onFragmentDestroyed() callback on Home Fragment *and* Login Fragment
- The postClear() resets the viewBinding to null again for the LoginFragment in your library

Later when an async operation is finished, I'm accessing the viewBinding again, but since it has become null, the library tries to recreate it, but fails with "view must have a tag" in the google databinding code (AGP 7.0.3)

`private inner class ClearOnDestroy : FragmentManager.FragmentLifecycleCallbacks() {

        override fun onFragmentDestroyed(fm: FragmentManager, f: Fragment) {
            // Fix for destroying view for case with issue of navigation
            postClear()
        }
    }`

I don't understand why the postClear() is called for the Login Fragment (onDestroyView() is not called by Android on that Login fragment). If it would not call postClear() I believe this issue would be fixed.

Now that I have a clearer understanding I can probably make a small reproducable testcase tomorrow.

**Comment by kirich1409** [comment]
I made fix for case when `Lifecycle.onDestroy()` wasn't called properly and I think this is the reason

**Comment by kirich1409** [comment]
@wbervoets , I've just published 1.5.3 and it must be available in few hours. Please check it and write is the fix working or not

**Comment by Stefanos65535** [comment]
same issue for my app
fixed by update to 1.5.3 
thx

**Comment by wbervoets** [comment]
yes, 1.5.3 works fine. Thanks!
